### ISSUE TITLE
AudioSession.instance creates new listeners or instance every time called on iOS

### ISSUE THREAD

**Issue by snaeji** [ryanheise/audio_session#6]
**Which API doesn't behave as documented, and how does it misbehave?**
Each time that we call `await AudioSession.instance` in the code or at startup `initWithRegistrar` is called.

This causes the following code being run every time and we end up with a lot of listeners of the same type running at the same time

```
    [AVAudioSession sharedInstance];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(audioInterrupt:) name:AVAudioSessionInterruptionNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(routeChange:) name:AVAudioSessionRouteChangeNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(silenceSecondaryAudio:) name:AVAudioSessionSilenceSecondaryAudioHintNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(mediaServicesLost:) name:AVAudioSessionMediaServicesWereLostNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(mediaServicesReset:) name:AVAudioSessionMediaServicesWereResetNotification object:nil];
```

**Minimal reproduction project**
Audio_service example project

**To Reproduce (i.e. user steps, not code)**
1. Press AudioPlayer
2. Press Stop
... repeate a few times

**Error messages**

We can see how the listeners stack up an we get a new listener for every start/stop

When a route change happens after 3 start/stops we get the following:
```
routeChange detected
routeChange detected
routeChange detected
routeChange detected
```

We can also put a NSLog print statement in `initWithRegistrar()` in `DarwinAudioSession.m`

**Expected behavior**
Each listener should only be called once

**Screenshots**

**Runtime Environment (please complete the following information if relevant):**
 - Device: Iphone Se (second generation)
 - OS: iOS 14.0.1

**Flutter SDK version**
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 1.22.0, on Mac OS X 10.15.7 19H2, locale en-GB)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
[✓] Android Studio (version 4.0)
[✓] IntelliJ IDEA Community Edition (version 2018.2.4)
[✓] VS Code (version 1.49.3)
[✓] Connected device (1 available)
    ! Error: Auglysingadeild4thgen is not connected. Xcode will continue when Auglysingadeild4thgen is
      connected. (code -13)
    ! Error: Living Room is not connected. Xcode will continue when Living Room is connected. (code
      -13)
    ! Error: Living Room is not connected. Xcode will continue when Living Room is connected. (code
      -13)

• No issues found!
```

**Comment by ryanheise** [comment]
Oh, you're right!

Both `addObserver` and `removeObserver` are tied to the plugin instance. What should happen is that `initWithRegistrar` and `dealloc` should both look at `sessions.count` and only call `addObserver` if this is the first session being created, and only call `removeObserver` if this is the last session being deallocated.

**Comment by ryanheise** [comment]
Not to mention that `dealloc` probably doesn't get called since each instance is retained in the `sessions` array. Probably the array of sessions needs to be `__weak`.

**Comment by snaeji** [comment]
Unfortunately my obj-c knowledge is very limited or almost non-existant so I have little there.... But I'm wondering what benefits there are of having multiple instances over one platform instance?

**Comment by ryanheise** [comment]
Plugins that do it that way generally break when they are loaded into apps that use more than one FlutterEngine instance (like audio_service-based apps, or any apps that do background execution).

Basically what should happen in an app that uses multiple FlutterEngines is that each FlutterEngine will have its own plugin registrar, and the plugins will be instantiated anew in each registrar. So the plugin should have an instance of itself loaded in each FlutterEngine, but it is up to the plugin to manage any parts of its state that are to be shared globally between all instances. You'll notice that the only instance data in `DarwinAudioSession` is the registrar and the channel, which need to be different per instance, while everything else was intended to be shared.

Technically `DarwinAudioSession` isn't itself the plugin, `AudioSessionPlugin` is, but the way I designed it is that `DarwinAudioSession` is kind of like a sub-plugin that behaves in the same way as a plugin. If I ever end up exposing another iOS API aside from `AVAudioSession`, the intention was that it would also be encapsulated in its own class, and itself behave in the same way as a plugin. That way, `AudioSessionPlugin` doesn't need to concern itself with how each individual API works.

**Comment by snaeji** [comment]
> Plugins that do it that way generally break when they are loaded into apps that use more than one FlutterEngine instance (like audio_service-based apps, or any apps that do background execution).
> 
> Basically what should happen in an app that uses multiple FlutterEngines is that each FlutterEngine will have its own plugin registrar, and the plugins will be instantiated anew in each registrar. So the plugin should have an instance of itself loaded in each FlutterEngine, but it is up to the plugin to manage any parts of its state that are to be shared globally between all instances. You'll notice that the only instance data in `DarwinAudioSession` is the registrar and the channel, which need to be different per instance, while everything else was intended to be shared.
> 
> Technically `DarwinAudioSession` isn't itself the plugin, `AudioSessionPlugin` is, but the way I designed it is that `DarwinAudioSession` is kind of like a sub-plugin that behaves in the same way as a plugin. If I ever end up exposing another iOS API aside from `AVAudioSession`, the intention was that it would also be encapsulated in its own class, and itself behave in the same way as a plugin. That way, `AudioSessionPlugin` doesn't need to concern itself with how each individual API works.

Thank you for taking the time to write this. It all makes sense to me now.

**Comment by ryanheise** [comment]
I have fixed the ARC management in the latest commit along with your issue, although fixing the ARC code has revealed a bug in just_audio and forced me to fix it's ARC code, too. I'll need to do some sort of coordinated release so that the new audio_session and just_audio are available at the same time.

I haven't pushed the just_audio fixes yet, so if you try the new audio_session with just_audio now, you'll experience a crash whenever a FlutterEngine is destroyed (when audio_service is stopped).

I might take a look at #7 before doing a release, and will let you know when it is available for testing.

**Comment by ryanheise** [comment]
Fixed in 0.0.9
